### PR TITLE
SarsaAlgorithm

### PR DESCRIPTION
**Pull request by emailweixu**
Tested on pendulum. In the figure, sac and ddpg is from sac_pendulum.gin and ddpg_pendulum.gin.
sarsa-legend
sarsa

### CODE REVIEW

File: alf/algorithms/sarsa_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            (time_step.observation, target_action),
+            step_type=time_step.step_type,
+            network_state=state.target_critic)
+
+        prev_return = tf.stop_gradient(time_step.reward +
+                                       time_step.discount * target_critic)
+        info = SarsaInfo(
+            actor_loss=actor_loss, critic=prev_critic, returns=prev_return)
+
+        if not is_train_step and not isinstance(self._actor_network,
```
Comment (witwolf) [inline alf/algorithms/sarsa_algorithm.py]: use `not isinstance(action_distribution, tf.Tensor)` like `common.to_distribution` seems better

File: alf/utils/value_ops.py [hunk 2]
```diff
@@ -194,10 +193,8 @@ def one_step_discounted_return(rewards, values, step_types, discounts):
     discounts = discounts[1:]
     rewards = rewards[1:]
     values = values[1:]
-    step_types = step_types[:-1]
 
-    is_lasts = tf.cast(tf.equal(step_types, StepType.LAST), tf.float32)
```
Comment (witwolf) [inline alf/utils/value_ops.py]: Why change this ? For `LAST step` , the value is  not  `reward + discount * V(FIRST step of new_episode)`,  it's `reward`. 
Thought it's masked out when calculate loss, it's better to use the correct calculation

File: alf/utils/metric_utils.py [hunk 3]
```diff
@@ -37,14 +37,17 @@ def __init__(self, time_step_spec, action_spec, policy_state_spec,
         super(Policy, self).__init__(
             time_step_spec=time_step_spec,
             action_spec=action_spec,
-            policy_state_spec=policy_state_spec)
+            policy_state_spec=(policy_state_spec, action_spec))
         # attribute named `_action_fn` already exist in parent
         self._action_fn1 = action_fn
 
     def _action(self, time_step, policy_state=(), seed=None):
-        policy_step = self._action_fn1(time_step, policy_state)
```
Comment (witwolf) [inline alf/utils/metric_utils.py]: `reset_state_if_necessary` for new episode .  And is there better way to store `prev_action` ?  it's very strange to store in `state`

File: alf/algorithms/sarsa_algorithm.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/sarsa_algorithm.py]: It's more trouble because action_distribution might be a nest.

File: alf/utils/value_ops.py (hunk 2 above)
Comment (emailweixu) [reply alf/utils/value_ops.py]: My bad. Will restore.

File: alf/utils/metric_utils.py (hunk 3 above)
Comment (emailweixu) [reply alf/utils/metric_utils.py]: tfa_eager_compute() handles that.

Comment (emailweixu) [reply alf/utils/metric_utils.py]: I agree. But this is the easiest way to add prev_action using tfa_eager_compute().
Perhaps we should use OnPolicyDriver for eval too.

File: alf/algorithms/sarsa_algorithm.py (hunk 1 above)
Comment (witwolf) [reply alf/algorithms/sarsa_algorithm.py]: i see

File: alf/algorithms/entropy_target_algorithm.py [hunk 4]
```diff
@@ -154,16 +157,43 @@ def train_step(self, distribution, step_type):
                 step_type=step_type,
                 loss=LossInfo(
                     loss=-entropy_for_gradient,
-                    extra=EntropyTargetLossInfo(entropy_loss=-entropy))))
+                    extra=EntropyTargetLossInfo(neg_entropy=-entropy))))
 
-    def calc_loss(self, training_info: EntropyTargetInfo):
+    def calc_loss(self, training_info: EntropyTargetInfo, valid_mask=None):
```
Comment (witwolf) [inline alf/algorithms/entropy_target_algorithm.py]: never pass  argument value for `valid_mask` when `calc_loss` is called,  and is it same with `mask` in L164 ?

Comment (emailweixu) [reply alf/algorithms/entropy_target_algorithm.py]: valid_mask will be used in some other places.
